tests/run/i1463.scala

```diff
@@ -0,0 +1,26 @@
+object Test {
+  case object Bob { override def equals(other: Any) = true }
+
+  class Bob2 {
+    override def equals(other: Any) = true
+  }
+  val Bob2 = new Bob2
+
+  def f0(x: Any) = x match { case Bob2 => Bob2 } // class cast exception at runtime, dotc only
+  def f1(x: Any) = x match { case Bob => Bob } // class cast exception at runtime, dotc only
+  def f2(x: Any): Bob.type = x match { case x @ Bob => x } // class cast exception at runtime, dotc and javac.
```

Comment:
Thanks @smarter , I see both approaches may cause some confusion to programmers. We need a spec here, are  you happy with what's in `-Xfuture` in 2.11 and 2.12, @odersky ?